src/nio/events/misc.py

```diff
@@ -114,9 +109,48 @@ class UnknownBadEvent:
     sender_key: Optional[str] = field(default=None, init=False)
     session_id: Optional[str] = field(default=None, init=False)
 
+    def flattened(
+        self,
+        _prefix: str = "",
+        _source: Optional[Dict[str, Any]] = None,
+        _flat: Optional[Dict[str, Any]] = None,
+    ) -> Dict[str, Any]:
+        """Return a flattened version of the ``source`` dict with dotted keys.
+
+        Example:
+            >>> event.source
+            {"content": {"body": "foo"}, "m.test": {"key": "bar"}}
+            >>> event.source.flattened()
+            {"content.body": "foo", "m.test.key": "bar"}
+
+        """
+
+        source = self.source if _source is None else _source
+        flat = {} if _flat is None else _flat
+
+        for key, value in source.items():
+            if isinstance(value, dict):
+                self.flattened(f"{_prefix}{key}.", value, flat)
+            else:
+                flat[f"{_prefix}{key}"] = value
+
+        return flat
+
+
+@dataclass
+class UnknownBadEvent(BaseEvent):
```

Comment:
I'm gonna have to think about this a bit when I'm back at a keyboard on whether this inheritance tree makes sense.

At least from a naming perspective, I think that this would probably make more sense to inherit from `BadEvent`, though

Reply:
I'm not sure that makes sense since `BadEvent` is a superset of `UnknownBadEvent`? you could do it the other way around and have `BadEvent` inherit from `UnknownBadEvent`, but then the names don't make sense, they'd have to be something like `KnownBadEvent` and `BadEvent`... except renaming would be a breaking change.